### PR TITLE
Add code to AAD provider and REGO for 7.6 through 7.9 to support PIM for Groups

### LINKED ISSUE #919: Add code to AAD provider and Rego for 7.6 through 7.9 to support PIM for Groups
# 💡 Summary #

Based on the prototyping performed for #792 we need to augment the existing AAD provider to extract the PIM for Groups configuration settings to support policies 7.6 through 7.9. Augment the Rego code to include the PIM for Groups configuration data in the policy evaluations.

## Implementation notes ##

Add code to the AAD provider Get-PrivilegedRole function:
- [ ] Whenever a group is found in the loop that iterates all the Active role assignments, check if that group is enrolled in PIM for Groups
- [ ] If the group is enrolled, then grab its configuration settings from PIM for Groups (get the "Member" config settings not the Owner)
- [ ] Leave the current code that grabs the configuration settings for the roles as-is. We still need to get the config data for the roles.
- [ ] Add code to load the PIM for Groups configurations into the JSON that is sent to the Rego. Each role can have one or more group configurations.

Add code to the AAD Rego for policies 7.6 through 7.9:

- [ ] In addition to checking the configurations for each role, the policies need to examine the configurations at the group level. For a policy to be compliant (Pass), the **role configurations AND the group configurations must be compliant** with the baseline policy.

*Note that in this new JSON structure that includes the PIM for Groups configurations, each role can have one more more group configurations associated with it.

### LINKED ISSUE #944: Add functional tests for PIM for Groups
# 💡 Summary #

Update AAD test plan to include test cases in 7.6 through 7.9 to cover PIM for Groups.

## Motivation and context ##

This would be useful because it will provide automated testing of this capability.

## Implementation notes ##

Please provide details for implementation, such as:

- Add compliant test case for each policy
- Add non-compliant test case for each policy 

## Acceptance criteria ##

How do we know when this work is done?

- [ ] New test case pass as expected

### LINKED ISSUE #975: MS Graph connection missing scope, RoleManagementPolicy.Read.AzureADGroup
# 🐛 Summary #

To support getting privileged role data Scuba users (including service principals) need elevated privileges.  

## To reproduce ##

1. Run nightly product tests. 

## Expected behavior ##

Expected aad tests to pass, failed due to lack of privileges

### LINKED ISSUE #981: AAD provider calls MS Graph Commands on tenants without proper license
# 🐛 Summary #

AAD provider calls MS Graph AAD PIM Commands on tenants without proper license.  Command, Get-MgBetaRoleManagementDirectoryRoleAssignmentScheduleInstance, should only be called if tenant has a premium license. 

## To reproduce ##

1. Run scuba against a tenant (i.e. g3) without a premium license (AAD P2)

## Expected behavior ##

Expected no PowerShell error (see policy 3.6 and 7.x)

## Any helpful log output or screenshots ##

Paste the results here:

```console

```

Add any screenshots of the problem here.

### PR DESCRIPTION
**Pull request by crutchfield**
#  #
  

## 🗣 Description ##

Updated AAD Provider to add PIM configurations to the JSON for groups that are enrolled in PIM for Groups in the Get-PrivilegedRole function. Modified the Rego to support reporting on PIM group configurations in addition to role configurations. This affects AAD policies 7.6 - 7.9. 

closes #919 
closes #981 
closes #975 
closes #944 

Salient items changed:

- Added code to the AAD provider Get-PrivilegedRole to support PIM for groups
- Modified Rego policies to report on roles and PIM groups
- Added required permissions to the Connection object
- Modified existing functional tests for 7.6 through 7.9 and added new tests for PIM for groups scenarios

## 💭 Motivation and context ##

PIM Groups provide mechanism to give users privileged access.  This enhancement allows to make sure notifications/alerts are produced appropriately. 

## 🧪 Testing ##

- Updated and ran REGO Unit test
- Created test scenarios on E5 and G5 tenants to test. Tested on G3 as well.
- Tested using the updated functional tests as well. 

## ✅ Pre-approval checklist ##

- [X] This PR has an informative and human-readable title.
- [X] PR targets the correct parent branch (e.g., main or release-name) for merge.
- [X] Changes are limited to a single goal - *eschew scope creep!*
- [X] Changes are sized such that they do not touch excessive number of files.
- [X] *All* future TODOs are captured in issues, which are referenced in code comments.
- [X] These code changes follow the ScubaGear [content style guide](https://github.com/cisagov/ScubaGear/blob/main/CONTENTSTYLEGUIDE.md).
- [X] Related issues these changes resolve are linked preferably via [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).
- [X] All relevant type-of-change labels added.
- [X] All relevant project fields are set.
- [X] All relevant repo and/or project documentation updated to reflect these changes.
- [X] Unit tests added/updated to cover PowerShell and Rego changes.
- [X] Functional tests added/updated to cover PowerShell and Rego changes. Except covered in #944 
- [X] All relevant functional tests passed.
- [X] All automated checks (e.g., linting, static analysis, unit/smoke tests) passed.

## ✅ Pre-merge checklist ##

- [ ] PR passed smoke test check.
- [ ] Feature branch has been rebased against changes from parent branch, as needed

  Use `Rebase branch` button below or use [this](https://www.digitalocean.com/community/tutorials/how-to-rebase-and-update-a-pull-request) reference to rebase from the command line.
- [ ] Resolved all merge conflicts on branch
- [ ] Notified merge coordinator that PR is ready for merge via comment mention

## ✅ Post-merge checklist ##

- [ ] Feature branch deleted after merge to clean up repository.
- [ ] Verified that all checks pass on parent branch (e.g., main or release-name) after merge.

### CODE REVIEW

**Review by ssatyapal123 — COMMENTED** [review]
Initial test on E5 looks good:

1. Created a new PIM Group (Shanti Test PIM Group)
2. Assigned a privileged role (Cloud Application Administrator) to the Group
3. Added a user as an Eligible member of the new group
4. Ran ScubaGear on AAD
5. Checked the ProviderSettingsExport.Json for the privileged_users and privileged_roles objects
6. Verified that the new role was listed under the user within the privileged_users array
7. Noted that the user UUID was not listed within the privileged_roles array under the Cloud Application Administrator grouping (Question - should it be listed here or should this section only list users assigned directly to the role and not through the PIM Group?)
8. Added the same user with an active assignment to the Cloud Application Administrator role
9. Ran ScubaGear on AAD
10. Reviewed the ProviderSettingsExport.json
11. Verified that the user UUID was now listed under Cloud Application Administrator privileged_roles grouping

Next step is to test G5 tenants and determine other ways of verifying the test passed besides reviewing the provider JSON.

**Review by tkol2022 — CHANGES_REQUESTED** [review]
Update the message for 7.6 to say "0 role(s) or group(s) allowing **activation** without approval found" without the bold.

image

**Comment by tkol2022** [comment]
### Testing Strategy
To test this PR, I'm using the following setup and expected result scenarios to exercise the potential permutations of compliant and non-compliant states. Each scenario assumes that a specific PIM group used for the test is **actively** assigned to one or more highly privileged roles - in the case of 7.6 the group is assigned to Global Admin.

**Scenario 1 Pass** - Both the role configurations and the PIM group configurations are compliant with the policy.
**Scenario 2 Fail** - The role configuration is compliant but the group configuration is non-compliant.
**Scenario 3 Fail** - The group configuration is compliant but the role configuration is non-compliant
**Scenario 4 Fail** - Both the role and the group configurations are non-compliant.

**Comment by tkol2022** [comment]
### Test Results

| Policy | Test results |
|---|---|
| 7.6 Activation of the Global Administrator role SHALL require approval. | Tested all four scenarios and the report displayed the correct output. |
| 7.7 Eligible and Active highly privileged role assignments SHALL trigger an alert. | col2 |
| 7.8 User activation of the Global Administrator role SHALL trigger an alert. | col2 |
| 7.9 User activation of other highly privileged roles SHOULD trigger an alert. | col2 |

**Comment by tkol2022** [comment]
### Regression Testing Results
Modified the test configurations for policies 7.4 and 7.5.
Ran ScubaGear version 1.1.1 versus the code in this PR to compare the outputs for those policies.
Policies 7.4 and 7.5 produce the same output for both versions. This matches what I expected.

**Comment by tkol2022** [comment]
### Code review of JSON
I reviewed the new provider JSON structure and observed that now the PIM group configurations were added to the respective roles where groups are assigned. The new fields Rule.RuleSource, Rule.RuleSourceType were added and referenced in the Rego. No questions about these specific updates.

image

**Comment by tkol2022** [comment]
> > Ran on G3 tenant and got the warning to terminal: "WARNING: Omitting calls to Get-PrivilegedRole and Get-PrivilegedUser."
> > The report showed "Error" in the Result column of the AAD-7 group policies. Wondering if there is a missing check to for tenants without the PIM capabilities.
> 
> can you post a screenshot? i will take a look
> 
> sure, here is a screenshot for G3:
> 
Shanti - Crutch is looking into why the code is giving errors for G3.

File: PowerShell/ScubaGear/Modules/Providers/ExportAADProvider.psm1 [hunk 1]
```diff
@@ -428,15 +482,13 @@ function Get-PrivilegedRole {
         $TenantHasPremiumLicense
     )
 
-    Write-Information "Entering"
     $PrivilegedRoles = [ScubaConfig]::ScubaDefault('DefaultPrivilegedRoles')
```
Comment (tkol2022) [inline PowerShell/ScubaGear/Modules/Providers/ExportAADProvider.psm1:485]: Crutch, can the config value for DefaultPrivilegedRoles be overridden by a custom config file created by the user?

**Comment by crutchfield** [comment]
> Update the message for 7.6 to say "0 role(s) or group(s) allowing **activation** without approval found" without the bold.
> 
> image

Inserted activation in details

**Comment by crutchfield** [comment]
> > > Ran on G3 tenant and got the warning to terminal: "WARNING: Omitting calls to Get-PrivilegedRole and Get-PrivilegedUser."
> > > The report showed "Error" in the Result column of the AAD-7 group policies. Wondering if there is a missing check to for tenants without the PIM capabilities.
> > 
> > 
> > can you post a screenshot? i will take a look
> > sure, here is a screenshot for G3:
> 
> Shanti - Crutch is looking into why the code is giving errors for G3.

Issue is directly related to expired license.  When retrieving Service Plans we check for provisioning status of 'Success'. In our case it is 'Disabled' since the license has expired.  Question is how to handle as was ask in another thread.  I believe this is out of scope of this issue.   

> > > Ran on G3 tenant and got the warning to terminal: "WARNING: Omitting calls to Get-PrivilegedRole and Get-PrivilegedUser."
> > > The report showed "Error" in the Result column of the AAD-7 group policies. Wondering if there is a missing check to for tenants without the PIM capabilities.
> > 
> > 
> > can you post a screenshot? i will take a look
> > sure, here is a screenshot for G3:
> 
> Shanti - Crutch is looking into why the code is giving errors for G3.

This is directly related to expired license.  We look for service plans with a provisioning status of 'Success'.  In this case it is 'Disabled'.  The code then indicate getting privileged roles & users will be omitted.  This begs the previously ask question, do we alert the user that the service plans are disabled?  I believe resolution is out of scope of this issue.  WIll open another issue to track.

**Review by ssatyapal123 — APPROVED** [review]
Ran against GCC High and verified results were as expected:

image

Also ran on E5, and verified results. Ran a scenario for 7.7 and verified that a role with a PIM assignment that was previously failing subsequently passed as expected.

**Comment by crutchfield** [comment]
@tkol2022  Added code to cache check for PIM Groups Looks like a good number of hits. With a large number of assignments the hit/miss ration would just get better.

File: PowerShell/ScubaGear/Modules/Providers/ExportAADProvider.psm1 [hunk 2]
```diff
@@ -336,37 +335,102 @@ function Get-PrivilegedUser {
...
+function IsPimGroup{
+    param(
+        [Parameter(Mandatory=$true)]
+        [ValidateNotNullOrEmpty()]
+        [string]
+        $GroupId
+    )
+
+    If ($null -eq [GroupType]::CheckedGroups[$GroupId]){
+        ($GroupEligibilitySchedule = Get-MgBetaIdentityGovernancePrivilegedAccessGroupEligibilitySchedule -Filter "groupId eq '$PrincipalId'") *> $null
```
Comment (crutchfield) [inline PowerShell/ScubaGear/Modules/Providers/ExportAADProvider.psm1:380]:
```suggestion
        ($GroupEligibilitySchedule = Get-MgBetaIdentityGovernancePrivilegedAccessGroupEligibilitySchedule -Filter "groupId eq '$GroupId'") *> $null
```

**Comment by crutchfield** [comment]
@tkol2022 I fixed IsPimGroup per our discussion

**Comment by crutchfield** [comment]
@tkol2022 Update caching to use privileged users functions determination of an Id as a user (i.e. non-group) to reduce future API calls in privileged role, in particular PIM Groups,  Also cleared cache after processing so would be clean if executed again.

**Comment by tkol2022** [comment]
This is blocked waiting the reactivation of the G3 tenant.

**Review by tkol2022 — APPROVED** [review]

**Comment by tkol2022** [comment]
@nanda-katikaneni ready for merge

**Pull request merged**